py/src/braintrust/logger.py

```diff
@@ -334,14 +347,25 @@ def __init__(self, name=None):
         except Exception:
             queue_size = 1000
         self.queue = queue.Queue(maxsize=queue_size)
+        # Each time we put items in the queue, we notify any waiting consumer
+        # threads to attempt a flush.
+        self.queue_filled_condition = threading.Condition()
 
         atexit.register(self._finalize)
 
     def log(self, *args):
         self._start()
         for event in args:
-            self.queue.put(event)
-        self.queue_filled_event.set()
+            while True:
+                try:
+                    self.queue.put_nowait(event)
+                    break
+                except queue.Full:
+                    # Notify consumers to start draining the queue.
+                    with self.queue_filled_condition:
+                        self.queue_filled_condition.notify()
```

Comment:
this seems a little dicey -- if i'm reading this correctly, we'll spin on one CPU until the queue clears up a bit. Is that correct?

Reply:
Yeah that's true. Maybe it makes sense to block until the queue clears up. I will adjust.